### PR TITLE
[SPARK-35737][SQL] Parse day-time interval literals to tightest types

### PR DESCRIPTION
**Pull request by sarutak**
### What changes were proposed in this pull request?

This PR add a feature which parse day-time interval literals to tightest type.

### Why are the changes needed?

To comply with the ANSI behavior.
For example, `INTERVAL '10 20:30' DAY TO MINUTE` should be parsed as `DayTimeIntervalType(DAY, MINUTE)` but not as `DayTimeIntervalType(DAY, SECOND)`.

### Does this PR introduce _any_ user-facing change?

No because `DayTimeIntervalType` will be introduced in `3.2.0`.

### How was this patch tested?

New tests.

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
cc @MaxGekk @yaooqinn FYI

**Comment by MaxGekk** [comment]
@sarutak We should support single field as well there:
https://github.com/apache/spark/blob/d53831ff5cef618cfbc08dcea84ca9bb61ed9903/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala#L2452-L2462
Please, open a sub-task in JIRA to don't forget about it if it is hard to do in this PR.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala [hunk 1]
```diff
@@ -2357,9 +2357,14 @@ class AstBuilder extends SqlBaseBaseVisitor[AnyRef] with SQLConfHelper with Logg
         Literal(calendarInterval.months, YearMonthIntervalType)
       } else {
         assert(calendarInterval.months == 0)
+        val strToFieldIndex = DayTimeIntervalType.dayTimeFields.map(i =>
```
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:2360]: I have seen this already in other PR. Does it make sense to put it to a common place `case object DayTimeIntervalType`?

**Review by MaxGekk — APPROVED** [review]
LGTM. Waiting for GA.

**Comment by sarutak** [comment]
> Please, open a sub-task in JIRA to don't forget about it if it is hard to do in this PR.

Yes, I've noticed that we should support single field but single field is currently parsed as `CalendarIntervalType` so we need change the behavior. So I separated that from this issue.
Sure, I'll open a sub task.

BTW, what do you think of multiple fields like `1 day 2 hours`? They should be parsed as `DayTimeIntervalType` rather than `CalendarType`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala (hunk 1 above)
Comment (sarutak) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:2360]: Yeah, it seems a common logic. After this and #32893 are merged, I'll open a followup PR.

**Comment by MaxGekk** [comment]
> BTW, what do you think of multiple fields like 1 day 2 hours? They should be parsed as DayTimeIntervalType rather than CalendarType?

I think we should take into account the SQL config `spark.sql.legacy.interval.enabled`. If it is false (default), we should parse `1 day 2 hours` as `DayTimeIntervalType`. There is a problem with mixed intervals like `1 month 2 hours`. In that case, we should throw an exception, and ask users either to change the config or interval (`1 month 2 hours` -> `30 days 2 hours`).

**Comment by MaxGekk** [comment]
+1, LGTM. Merging to master.
Thank you, @sarutak .

**Pull request closed without merging**